spring-cloud-dataflow-core/src/main/java/org/springframework/cloud/dataflow/core/dsl/AbstractTokenizer.java

```diff
@@ -0,0 +1,330 @@
+ */
+public abstract class AbstractTokenizer {
+
+	private final int[] NO_LINEBREAKS = new int[0];
+	
+	/**
+	 * The string to be tokenized.
+	 */
+	protected String expressionString;
+
+	/**
+	 * The expressionString as a char array.
+	 */
+	protected char[] toProcess;
+
+	/**
+	 * Length of input data.
+	 */
+	protected int max;
+
+	/**
+	 * Current lexing position in the input data.
+	 */
+	protected int pos;
+
+	/**
+	 * Output stream of tokens.
+	 */
+	protected List<Token> tokens = new ArrayList<Token>();
+	
+	/**
+	 * Positions of linebreaks in the parsed string.
+	 */
+	protected int[] linebreaks = NO_LINEBREAKS;
+
+	abstract void process();
+	
+	public Tokens getTokens(String inputData) {
+		this.expressionString = inputData;
+		this.toProcess = (inputData + "\0").toCharArray();
+		this.max = toProcess.length;
+		this.pos = 0;
+		this.tokens.clear();
+		process();
+		return new Tokens(inputData, tokens, linebreaks);
+	}
+
+	/**
+	 * Check if this might be a two character token.
+	 */
+	protected boolean isTwoCharToken(TokenKind kind) {
+		Assert.isTrue(kind.tokenChars.length == 2);
+		Assert.isTrue(toProcess[pos] == kind.tokenChars[0]);
+		return toProcess[pos + 1] == kind.tokenChars[1];
+	}
+
+	/**
+	 * Push a token of just one character in length.
+	 */
+	protected void pushCharToken(TokenKind kind) {
+		tokens.add(new Token(kind, pos, pos + 1));
+		pos++;
+	}
+
+	/**
+	 * Push a token of two characters in length.
+	 */
+	protected void pushPairToken(TokenKind kind) {
+		tokens.add(new Token(kind, pos, pos + 2));
+		pos += 2;
+	}
+
+	protected boolean isIdentifier(char ch) {
+		return isAlphabetic(ch) || isDigit(ch) || ch == '_' || ch == '$' || ch == '-';
+	}
+
+	protected boolean isQuote(char ch) {
+		return ch == '\'' || ch == '"';
+	}
+
+	protected boolean isWhitespace(char ch) {
+		return ch == ' ' || ch == '\t' || ch == '\r' || ch == '\n';
+	}
+
+	protected boolean isDigit(char ch) {
+		if (ch > 255) {
+			return false;
+		}
+		return (flags[ch] & IS_DIGIT) != 0;
+	}
+
+	protected boolean isAlphabetic(char ch) {
+		if (ch > 255) {
+			return false;
+		}
+		return (flags[ch] & IS_ALPHA) != 0;
+	}
+
+	protected char[] subArray(int start, int end) {
+		char[] result = new char[end - start];
+		System.arraycopy(toProcess, start, result, 0, end - start);
+		return result;
+	}
+
+	protected void lexIdentifier() {
+		int start = pos;
+		do {
+			pos++;
+		}
+		while (isIdentifier(toProcess[pos]));
+		char[] subarray = subArray(start, pos);
+		tokens.add(new Token(TokenKind.IDENTIFIER, subarray, start, pos));
+	}
+
+	/**
+	 * Lex a string literal which uses single quotes as delimiters. To include
+	 * a single quote within the literal, use a pair ''
+	 */
+	protected void lexQuotedStringLiteral() {
+		int start = pos;
+		boolean terminated = false;
+		while (!terminated) {
+			pos++;
+			char ch = toProcess[pos];
+			if (ch == '\'') {
+				// may not be the end if the char after is also a '
+				if (toProcess[pos + 1] == '\'') {
```

Comment:
We don't nicely handle them doing their own escaping. If they want to include a ' inside the string then they provide a double '' and it is treated as a single one. This isn't new, it is what stream tokenization always did - although I can't quite remember where that notion came from! (This AbstractTokenizer is mostly the old stream tokenizer).